src/framework.js

```diff
@@ -0,0 +1,194 @@
+/**
+ * File: framework.js 
+ */
+var _ = require('underscore');
+var util = require('util');
+
+var stateFactory = require('./create-state');
+var vnis = require('./vnis');
+
+/**
+ * This module implements the core of the RDF Pipeline Framework.  
+ */
+module.exports = {
+
+    componentFactory: function( nodeDef, wrapper ) { 
+        
+        var compFactory = require('./noflo-component-factory');
+        return compFactory(
+                _.defaults( {
+                     // Add an ondata function to be invoked every time we receive new data
+                     inPorts: _.mapObject(nodeDef.inPorts, function( port, portName ) {  
+                         return _.extend({
+                             ondata: function(payload, socketIndex) {
+                                   var that = this; 
+                                   return ondata.call( that, payload, portName, socketIndex, port );
+                             }
+                         }, port);
+                     })
+                 }, nodeDef ),
+                { rpf: {      // extend default noflo component node instance with framework rpf object 
+                     vnis: vnis,
+                     wrapper: wrapper
+                  }
+                } 
+        );
+    }
+
+}; // module.exports
+
+
+var ondata = function( payload, portName, socketIndex, port ) {
+
+     // socketIndex is only defined for addressable/multi ports so set default of 0 if not defined.  
+     // If there is only one input, we can just default to socket index 0, so if not defined, 
+     // do this for now.   We need to discuss whether addressable should just be required on our 
+     // pipeline ports.  IMHO, I think it probably should be to minimize potential user-error in this area
+     // and simplify the interface our users work with.
```

Comment:
The policy that I planned was: 1. users must set the 'multi' option to true if they want to attach more than one thing to the port; and 2. an array of data objects (probably {sourceNodeName: ssn, sourcePortName: spn, data: data} objects) is always passed to each updater parameter whose port was declared 'multi'.  Some consequences of this policy: (a) it's an error if more than one thing is attached to a node that was not declared 'multi'; (b) if a port is 'required' and 'multi' then it is an error if it has zero things attached; (c) if a port is 'multi' and not 'required' and has zero things attached then its corresponding updater parameter receives an array with zero elements.